### ISSUE TITLE
Error 451 with https://ghapi.huchen.dev/

### ISSUE THREAD

**Issue by hedythedev** [huchenme/github-trending-api#130]
![image](https://user-images.githubusercontent.com/50042066/95010270-797d2500-065a-11eb-9cf2-7e6bb0a786dc.png)

**Comment by nahtnam** [comment]
Same here

**Comment by huchenme** [comment]
Unfortunately Vercel did not want to sponsor this project as it is API only. I could no longer find a good free host for the project. I suggest you clone the repo and deploy the code somewhere 😞

**Comment by omarryhan** [comment]
That's sad news :/

@github should definitely be sponsoring this or atleast provide an API of their own.

**Comment by omarryhan** [comment]
@huchenme How about we add a simple frontend that makes use of the API, that wouldn't be breaching Vercel's TOS, right?

**Comment by huchenme** [comment]
I can try add frontend for this actually

**Comment by sandoche** [comment]
I can give you the code of GitNews (git.news) if you wish.

**Comment by huchenme** [comment]
@sandoche I do have some code, let me add to the repo during nights and apply Vercel's sponsor again, if everything goes smooth it should be back in a week 🤞 
Meanwhile, if you need the API now, I suggest you clone the code and deploy to somewhere like Vercel

**Comment by akane10** [comment]
I use this api for my personal website. Since it's having an issue, I cloned and hosted it. 
So in the meantime, feel free to use it, `gtrend.yapie.me`

**Comment by huchenme** [comment]
@akane10 that is great, may I know where did you host it?

**Comment by akane10** [comment]
@huchenme yeah sure, it's on digitalocean.
But anyway, isn't there netlify function and it is free?

**Comment by huchenme** [comment]
@akane10 I tried Netlify function and could not recall what did not work as expected. if you plan to keep your server, do you mind if I point `ghapi.huchen.dev` to your server temporarily?

**Comment by akane10** [comment]
@huchenme Yes, sure. I don't mind

**Comment by Smeet97Kathiria** [comment]
did you try it using this method for netlify ? @huchenme  - https://docs.netlify.com/site-deploys/create-deploys/#deploy-with-git

**Comment by huchenme** [comment]
The issue I was facing for Netlify is the caching: `s-maxage` does not actually work: if I hit `https://ghapi.huchen.dev/repositories?language=javascript`, all the subsequent requests with different parameters (e.g. https://ghapi.huchen.dev/repositories?language=java) will return the same results (javascript)

If anyone has any luck with Netlify please let me know

**Comment by huchenme** [comment]
This is miracle.
![image](https://user-images.githubusercontent.com/2078389/95215314-6ec1bc00-0823-11eb-82d9-6e022fb9c1ee.png)

I am thinking to deploy to Netlify without caching. It might be slower, and it might hit the limit soon as well 🤞

**Comment by huchenme** [comment]
It is back now! the only downside is that it does not have browser cache now

**Comment by huchenme** [comment]
It quickly finished the free quota in Netlify, have to stop the project.

**Comment by nahtnam** [comment]
Hmm, have you considered instead making a library that we can import and await and it will make the request and parse the request for us? That way you don't have to worry about hosting it, and we are able to use the code you wrote

**Comment by akane10** [comment]
@nahtnam he already did it actually `https://github.com/huchenme/github-trending-api#install`.
And yeah I agree, that's totally fine if there is no API, only library. 
We appreciate your work @huchenme.

**Comment by pupubird** [comment]
Hey hey for anyone that's reading on this, I had been using this library for quite a long time, definitely appreciate your work! @huchenme !

I had also created a self-hosted version of the scraper server ( with AWS lambda function compatible ), [https://github.com/pupubird/get-github-trending](https://github.com/pupubird/get-github-trending)

And the demo site here [https://hackertab.pupubird.com/repositories](https://hackertab.pupubird.com/repositories)

All credits go to @huchenme

**Comment by pupubird** [comment]
Also, by utilizing AWS CloudFront, I am able to cache the response with TTL to daily, will write up a tutorial about it too!

**Comment by waningflow** [comment]
Feel free to use.
https://github-trending-api.waningflow.com/languages
https://github-trending-api.waningflow.com/repositories?language=typescript&since=daily

**Comment by Smeet97Kathiria** [comment]
@waningflow will this ever go down?

**Comment by waningflow** [comment]
@Smeet97Kathiria Deployed on Function Compute of Aliyun. About first 500k requests per month are free.

**Comment by Smeet97Kathiria** [comment]
@waningflow  what happens after it hit the limits?

**Comment by waningflow** [comment]
@Smeet97Kathiria I suppose it's not likely to reach the limit for normal usage. If it does, I may restrict the apis temprorily. It's alse recommended to deploy the service youself. Here is what I use in Function Compute of Aliyun, [github-trending-api-node](https://github.com/waningflow/github-trending-api-node). Just `yarn` to install dependencies and upload the whole git to the platform.

**Comment by hedythedev** [comment]
hey @akane10 I tried out your url, and it seems https://gtrend.yapie.me/spoken_languages doesn't work?

**Comment by akane10** [comment]
@hedythedev yeah, route `languages` and `spoken_languages` was I disabled for a reason. 
But it is now working. 
In case anyone uses my endpoints, it won't run out of request, as long as I can affort my VM.

**Comment by andraz-at** [comment]
Can you workaround this issue by creating an AWS hosted endpoint (trivially built with AWS Amplify -> Hosting)?

Then you just need to add a Lambda function which calculates all required responses and stores the responses into static files to an S3 bucket. From this bucket you can serve static files as a web page by simply enabling public access to this static hosting (amplify hosting does that in the background for you).


tl;dr: Create a trivial "web site" to set up free web hosting + create a Lambda function to update the same "web site's" files in the S3 bucket (fetch, parse, write json...).

**Comment by omarryhan** [comment]
@andraz-at Instead of making the function store the response on S3, wouldn't it be simpler to cache all the function responses on a CDN e.g. AWS Cloudfront or Cloudflare? Serving from a CDN cache is even cheaper than serving directly from S3.

Since these responses will only change once a day, you can set the TTL to be 86400 seconds (1 day). That way, it's very unlikely you run out of free function calls on most FAAS providers.

If my memory serves me correctly, @huchenme was actually caching the responses when this project was being hosted on Vercel, it's a bummer that they suspended this project and his account as well :/

**Comment by cm3z4** [comment]
Hey @akane10, I'm currently routing my site/app [GitYeti](http://gityeti.herokuapp.com/l) to your server. Is that okay?

**Comment by pupubird** [comment]
> Hey hey for anyone that's reading on this, I had been using this library for quite a long time, definitely appreciate your work! @huchenme !
> 
> I had also created a self-hosted version of the scraper server ( with AWS lambda function compatible ), https://github.com/pupubird/get-github-trending
> 
> And the demo site here https://hackertab.pupubird.com/repositories
> 
> All credits go to @huchenme

@andraz-at hey, I had already implemented caching on CloudFront site ( the cdn ), do feel free to directly deploy to your aws account using `serverless deploy` !

**Comment by akane10** [comment]
> Hey @akane10, I'm currently routing my site/app [GitYeti](http://gityeti.herokuapp.com/l) to your server. Is that okay?

@cm3z4 yes, sure. Feel free to use it.

**Comment by wonderbeyond** [comment]
@huchenme 
I've tried hard to revert your recent commits which are tightly bound to now.sh and netlify.
I recalled the `src/server.js` file to make it able to run as a normal web server.
Then I made a deployment to heroku: https://github-trending-api-wonder.herokuapp.com/

My fork: https://github.com/wonderbeyond/github-trending-api/

**Comment by andraz-at** [comment]
> @andraz-at Instead of making the function store the response on S3, wouldn't it be simpler to cache all the function responses on a CDN e.g. AWS Cloudfront or Cloudflare? Serving from a CDN cache is even cheaper than serving directly from S3.
> 
> Since these responses will only change once a day, you can set the TTL to be 86400 seconds (1 day). That way, it's very unlikely you run out of free function calls on most FAAS providers.

Yes, there are alternative optimizations available. My point was that we can chain all those optimizations one after another in a way to fit all requests into the free tier limit of AWS (preferably the permanent one, not the 12 months one).

Alternative would be to use AWS lambda periodically (CRON event) only for processing the data, then push the hosting on GitHub pages / GitHub gists by programmatically updating those files remotely. This would also be a nice way of making GitHub host their own feed by serving static files. 🙂 

We can also do both, to spread the load. But premature optimization is a slippery slope to stop doing anything at all, so we better skip digging deeper if not needed.

**Comment by Smeet97Kathiria** [comment]
@waningflow are both of your urls down?

**Comment by Smeet97Kathiria** [comment]
HIi @akane10 I am currently routing my app https://apps.apple.com/us/app/superrepo/id1517331914 to your server. is that okay?

**Comment by Smeet97Kathiria** [comment]
Hi, @pupubird can I route my app https://apps.apple.com/us/app/superrepo/id1517331914 to your site?

**Comment by pupubird** [comment]
Hey @Smeet97Kathiria sure!

**Comment by pupubird** [comment]
Cuz I am using it for the unofficial hackertab https://github.com/huchenme/hacker-tab-extension/issues/55 too, so you are free to use as well

**Comment by Smeet97Kathiria** [comment]
@pupubird Thanks

**Comment by pupubird** [comment]
welcome @Smeet97Kathiria

**Comment by yvonmanzi** [comment]
> Hey hey for anyone that's reading on this, I had been using this library for quite a long time, definitely appreciate your work! @huchenme !
> 
> I had also created a self-hosted version of the scraper server ( with AWS lambda function compatible ), https://github.com/pupubird/get-github-trending
> 
> And the demo site here https://hackertab.pupubird.com/repositories
> 
> All credits go to @huchenme

This looks pretty good! Good work!

**Comment by RishabhKodes** [comment]
Guys feel free to use mine! [http://192.168.1.33:8000/](http://192.168.1.33:8000/)

Thanks

**Comment by Sky24n** [comment]
> 
> 
> Feel free to use.
> https://github-trending-api.waningflow.com/languages
> https://github-trending-api.waningflow.com/repositories?language=typescript&since=daily



> 
> 
> Feel free to use.
> https://github-trending-api.waningflow.com/languages
> https://github-trending-api.waningflow.com/repositories?language=typescript&since=daily

Don't you support developers?
https://github-trending-api.waningflow.com/developers?language=typescript&since=daily

**Comment by Smeet97Kathiria** [comment]
@wonderbeyond Can I route my app https://apps.apple.com/us/app/superrepo/id1517331914  to your server?
